File: src/RadioConfig.cpp
/*
 * mm-radio HAL (https://github.com/GloDroid/mm-radio)
 *
 * Copyright (C) 2021 The Android Open Source Project
 * Copyright (C) 2022 GloDroid project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "RadioConfig.h"

#include <aidl/android/hardware/radio/sim/BnRadioSim.h>

#include "debug.h"

#define RADIO_MODULE "Config"

namespace android::hardware::radio::mm {

using ::aidl::android::hardware::radio::RadioIndicationType;
using ::ndk::ScopedAStatus;
namespace aidl = ::aidl::android::hardware::radio::config;
constexpr auto ok = &ScopedAStatus::ok;

ScopedAStatus RadioConfig::getHalDeviceCapabilities(int32_t serial) {
    LOG_CALL << serial;
    mResponse->getHalDeviceCapabilitiesResponse(okay(serial), true);
    return ok();
}

ScopedAStatus RadioConfig::getNumOfLiveModems(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->getNumOfLiveModemsResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioConfig::getPhoneCapability(int32_t serial) {
    LOG_STUB << serial;
    auto cap = (aidl::PhoneCapability){
            .maxActiveData = 1,
            .maxActiveInternetData = 1,
            .isInternetLingeringSupported = false,
            .logicalModemIds = std::vector<uint8_t>(0),
    };

    mResponse->getPhoneCapabilityResponse(okay(serial), cap);

    return ok();
}

ScopedAStatus RadioConfig::getSimSlotsStatus(int32_t serial) {
    LOG_STUB << serial;
    mResponse->getSimSlotsStatusResponse(okay(serial), mGetSimSlotStatusCb());
    return ok();
}

ScopedAStatus RadioConfig::setNumOfLiveModems(int32_t serial, int8_t /*numOfLiveModems*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setNumOfLiveModemsResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioConfig::setPreferredDataModem(int32_t serial, int8_t modemId) {
    LOG_STUB << serial << " modemId: " << (int)modemId;
    auto resp = okay(serial);
    if (modemId < 0) resp.error = RadioError::INVALID_ARGUMENTS;
    mResponse->setPreferredDataModemResponse(resp);
    return ok();
}

ScopedAStatus RadioConfig::setResponseFunctions(
        const std::shared_ptr<aidl::IRadioConfigResponse>& radioConfigResponse,
        const std::shared_ptr<aidl::IRadioConfigIndication>& radioConfigIndication) {
    LOG_CALL << radioConfigResponse << ' ' << radioConfigIndication;

    CHECK(radioConfigResponse);
    CHECK(radioConfigIndication);

    mResponse = radioConfigResponse;
    mIndication = radioConfigIndication;

    return ok();
}

ScopedAStatus RadioConfig::setSimSlotsMapping(  //
        int32_t serial, const std::vector<aidl::SlotPortMapping>& /*slotMap*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setSimSlotsMappingResponse(okay(serial));
    return ok();
}

// Internal API:
void RadioConfig::simSlotStatusChanged() {
    LOG_CALL;
    if (mIndication)
        mIndication->simSlotsStatusChanged(RadioIndicationType::UNSOLICITED, mGetSimSlotStatusCb());
}

}  // namespace android::hardware::radio::mm

File: src/radio/radio_config.rs
/*
 * mm-radio HAL (https://github.com/GloDroid/mm-radio)
 *
 * SPDX-License-Identifier: GPL-3.0-only
 * Copyright (C) 2023 The GloDroid Project
 */

use crate::utils::iradio::{
    declare_async_iradio, entry_check, invalid_arg, not_implemented, okay, resp, resp_ok, shared,
    sharedmut,
};
use android_hardware_radio::aidl::android::hardware::radio::RadioIndicationType::*;
use android_hardware_radio_config::aidl::android::hardware::radio::config::{
    IRadioConfig::*, IRadioConfigIndication::*, IRadioConfigResponse::*, PhoneCapability::*,
    SimPortInfo::*, SimSlotStatus::*, SlotPortMapping::*,
};
use android_hardware_radio_sim::aidl::android::hardware::radio::sim::CardStatus::STATE_PRESENT;
use async_std::sync::RwLock;
use binder::{BinderFeatures, Strong};
use log::info;
use std::sync::Arc;
use zbus::export::async_trait::async_trait;

#[derive(Default)]
pub struct RadioConfigShared {
    /* modem_bound is used here only for unification of macros. Should be always true! */
    pub modem_bound: bool,
    response: Option<binder::Strong<dyn IRadioConfigResponse>>,
    indication: Option<binder::Strong<dyn IRadioConfigIndication>>,

    slot_status: Vec<SimSlotStatus>,
    slot_mapping: Vec<SlotPortMapping>,
}

#[derive(Default)]
pub struct RadioConfig {
    shared: Arc<RwLock<RadioConfigShared>>,
}

#[async_trait]
impl IRadioConfigAsyncServer for RadioConfig {
    async fn getSimSlotsStatus(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getSimSlotsStatusResponse, Default::default());
        let slot_status = shared!(&self).slot_status.clone();
        okay!(&self, serial, getSimSlotsStatusResponse, &slot_status)
    }
    async fn setSimSlotsMapping(
        &self,
        serial: i32,
        smap: &[SlotPortMapping],
    ) -> binder::Result<()> {
        entry_check!(&self, serial, setSimSlotsMappingResponse);
        sharedmut!(&self).slot_mapping = smap.to_vec();
        okay!(&self, serial, setSimSlotsMappingResponse)
    }
    async fn getHalDeviceCapabilities(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getHalDeviceCapabilitiesResponse, false);
        resp!(&self).getHalDeviceCapabilitiesResponse(&resp_ok(serial), true)
    }
    async fn getNumOfLiveModems(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getNumOfLiveModemsResponse, 0);
        resp!(&self).getNumOfLiveModemsResponse(&resp_ok(serial), 0)
    }
    async fn getPhoneCapability(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getPhoneCapabilityResponse, &Default::default());
        let pc = PhoneCapability {
            isInternetLingeringSupported: false,
            maxActiveData: 0,
            maxActiveInternetData: 0,
            logicalModemIds: vec![0],
        };
        resp!(&self).getPhoneCapabilityResponse(&resp_ok(serial), &pc)
    }
    async fn setNumOfLiveModems(&self, serial: i32, _num_of_live_modems: i8) -> binder::Result<()> {
        not_implemented!(&self, serial, setNumOfLiveModemsResponse)
    }
    async fn setPreferredDataModem(&self, serial: i32, modem_id: i8) -> binder::Result<()> {
        if modem_id < 0 {
            return invalid_arg!(&self, serial, setPreferredDataModemResponse);
        }
        entry_check!(&self, serial, setPreferredDataModemResponse);
        info!("setPreferredDataModem: modem = {}", modem_id);
        okay!(&self, serial, setPreferredDataModemResponse)
    }
    async fn setResponseFunctions(
        &self,
        radio_response: &binder::Strong<dyn IRadioConfigResponse>,
        radio_indication: &binder::Strong<dyn IRadioConfigIndication>,
    ) -> binder::Result<()> {
        info!("RadioConfig::setResponseFunctions");
        let mut shared = sharedmut!(&self);
        shared.response = Some(radio_response.clone());
        shared.indication = Some(radio_indication.clone());
        shared.modem_bound = true; /* for compatibility */
        shared.slot_status = vec![SimSlotStatus {
            cardState: STATE_PRESENT,
            portInfo: vec![SimPortInfo {
                iccId: "8938003992840681512F".to_string(),
                logicalSlotId: 0,
                portActive: true,
            }],
            ..Default::default()
        }];
        radio_indication
            .simSlotsStatusChanged(RadioIndicationType::UNSOLICITED, &shared.slot_status)?;

        Ok(())
    }
}

declare_async_iradio!(RadioConfig, IRadioConfig, BnRadioConfig);
